File: views/programming/c++/code-golf/GenomicRangeQuery.cpp
// Link to test: https://codility.com/demo/take-sample-test/genomic_range_query/
// "Find the minimal nucleotide from a range of sequence DNA."

// 62/100 solution. Produces correct output, but fails performance tests. I need a time complexity of O(N+M).
vector<int> genomic_range_query(string &S, vector<int> &P, vector<int> &Q){
    vector<int> low(P.size());
    for(unsigned int i = 0; i < P.size(); i++){
        for(int j = P[i]; j <= Q[i]; j++){
            switch (S[j]){
                case 'A':
                    S[j] = 1;
                    break;
                case 'C':
                    S[j] = 2;
                    break;
                case 'G':
                    S[j] = 3;
                    break;
                case 'T':
                    S[j] = 4;
                    break;
            }
            if(S[j] < low[i] || low[i] == 0) low[i] = S[j];
        }
    }
    return low;
}

File: views/programming/php/code-golf/GenomicRangeQuery.php
// Link to test: https://codility.com/demo/take-sample-test/genomic_range_query/
// "Count minimal number of jumps from position X to Y."

function genomic_range_query($S, $P, $Q) {
    for($i = 0; $i < sizeof($P);$i++){
        $temp = substr($S, $P[$i], $Q[$i] - $P[$i] + 1);
        if(strpos($temp, 'A') !== false){
            $min[$i] = 1;
        }elseif(strpos($temp, 'C') !== false){
            $min[$i] = 2;
        }elseif(strpos($temp, 'G') !== false){
            $min[$i] = 3;
        }elseif(strpos($temp, 'T') !== false){
            $min[$i] = 4;
        }
    }
    return $min;
}